File: GoogleCheckout.Source/examples/vb.net/api25notification/GoogleCheckoutHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports GCheckout.OrderProcessing
Imports GCheckout.Util

'original source from http://www.capprime.com/software_development_weblog/2010/11/29/UsingTheGoogleCheckout25APIWithASPNETMVCTheMissingSample.aspx

Public Class GoogleCheckoutHelper
    ' Methods
    Private Shared Sub HandleAuthorizationAmountNotification(ByVal inputAuthorizationAmountNotification As GCheckout.AutoGen.AuthorizationAmountNotification)
        'TODO: Add custom processing for this notification type
    End Sub

    Private Shared Sub HandleChargeAmountNotification(ByVal inputChargeAmountNotification As GCheckout.AutoGen.ChargeAmountNotification)
        'TODO: Add custom processing for this notification type
    End Sub

    Private Shared Sub HandleNewOrderNotification(ByVal inputNewOrderNotification As GCheckout.AutoGen.NewOrderNotification)
        Dim hiddenMerchantPrivateData As String = inputNewOrderNotification.shoppingcart.merchantprivatedata.Any(0).InnerText
        'TODO: Process the MerchantPrivateData if provided

        For Each oneItem As GCheckout.AutoGen.Item In inputNewOrderNotification.shoppingcart.items
            ' TODO: Get MerchantItemId from shopping cart item (oneItem.merchantitemid) and process it
        Next

        'TODO: Add custom processing for this notification type

    End Sub

    Private Shared Sub HandleOrderStateChangeNotification(ByVal notification As GCheckout.AutoGen.OrderStateChangeNotification)

        'Charge Order If Chargeable
        If ((notification.previousfinancialorderstate = GCheckout.AutoGen.FinancialOrderState.REVIEWING) _
            AndAlso (notification.newfinancialorderstate = GCheckout.AutoGen.FinancialOrderState.CHARGEABLE)) Then

            Dim oneGCheckoutResponse As GCheckoutResponse = New ChargeOrderRequest(notification.googleordernumber).Send
        End If
        'Update License If Charged
        If ((notification.previousfinancialorderstate = GCheckout.AutoGen.FinancialOrderState.CHARGING) _
            AndAlso (notification.newfinancialorderstate = GCheckout.AutoGen.FinancialOrderState.CHARGED)) Then
            'TODO: For each shopping cart item received in the NewOrderNotification, authorize the license
        End If

        'TODO: Add custom processing for this notification type
    End Sub

    Private Shared Sub HandleRiskInformationNotification(ByVal notification As GCheckout.AutoGen.RiskInformationNotification)
        ' TODO: Add custom processing for this notification type
    End Sub

    Public Shared Sub ProcessNotification(ByVal serialNumber As String)

        'The next two statements set up a request and call google checkout for the details based on that serial number.
        Dim oneNotificationHistoryRequest As New NotificationHistoryRequest(New NotificationHistorySerialNumber(serialNumber))
        Dim oneNotificationHistoryResponse As NotificationHistoryResponse = DirectCast(oneNotificationHistoryRequest.Send, NotificationHistoryResponse)

        'oneNotificationHistoryResponse.ResponseXml contains the complete response

        Dim oneNotification As Object
        For Each oneNotification In oneNotificationHistoryResponse.NotificationResponses
            If oneNotification.GetType.Equals(GetType(GCheckout.AutoGen.NewOrderNotification)) Then
                Dim oneNewOrderNotification As GCheckout.AutoGen.NewOrderNotification = DirectCast(oneNotification, GCheckout.AutoGen.NewOrderNotification)
                If oneNewOrderNotification.serialnumber.Equals(serialNumber) Then
                    GoogleCheckoutHelper.HandleNewOrderNotification(oneNewOrderNotification)
                End If
            ElseIf oneNotification.GetType.Equals(GetType(GCheckout.AutoGen.OrderStateChangeNotification)) Then
                Dim oneOrderStateChangeNotification As GCheckout.AutoGen.OrderStateChangeNotification = DirectCast(oneNotification, GCheckout.AutoGen.OrderStateChangeNotification)
                If oneOrderStateChangeNotification.serialnumber.Equals(serialNumber) Then
                    GoogleCheckoutHelper.HandleOrderStateChangeNotification(oneOrderStateChangeNotification)
                End If
            ElseIf oneNotification.GetType.Equals(GetType(GCheckout.AutoGen.RiskInformationNotification)) Then
                Dim oneRiskInformationNotification As GCheckout.AutoGen.RiskInformationNotification = DirectCast(oneNotification, GCheckout.AutoGen.RiskInformationNotification)
                If oneRiskInformationNotification.serialnumber.Equals(serialNumber) Then
                    GoogleCheckoutHelper.HandleRiskInformationNotification(oneRiskInformationNotification)
                End If
            ElseIf oneNotification.GetType.Equals(GetType(GCheckout.AutoGen.AuthorizationAmountNotification)) Then
                Dim oneAuthorizationAmountNotification As GCheckout.AutoGen.AuthorizationAmountNotification = DirectCast(oneNotification, GCheckout.AutoGen.AuthorizationAmountNotification)
                If oneAuthorizationAmountNotification.serialnumber.Equals(serialNumber) Then
                    GoogleCheckoutHelper.HandleAuthorizationAmountNotification(oneAuthorizationAmountNotification)
                End If
            ElseIf oneNotification.GetType.Equals(GetType(GCheckout.AutoGen.ChargeAmountNotification)) Then
                Dim oneChargeAmountNotification As GCheckout.AutoGen.ChargeAmountNotification = DirectCast(oneNotification, GCheckout.AutoGen.ChargeAmountNotification)
                If oneChargeAmountNotification.serialnumber.Equals(serialNumber) Then
                    GoogleCheckoutHelper.HandleChargeAmountNotification(oneChargeAmountNotification)
                End If
            Else
                Throw New ArgumentOutOfRangeException(String.Concat(New String() {"Unhandled Type [", oneNotification.GetType.ToString, "]!; serialNumber=[", serialNumber, "];"}))
            End If
        Next
    End Sub


End Class

File: GoogleCheckout.Source/examples/api25notification/App_Code/GoogleCheckoutHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using GCheckout.OrderProcessing;

//original source from http://www.capprime.com/software_development_weblog/2010/11/29/UsingTheGoogleCheckout25APIWithASPNETMVCTheMissingSample.aspx

/// <summary>
/// Summary description for GoogleCheckoutHelper
/// </summary>
public class GoogleCheckoutHelper {

  private static void HandleAuthorizationAmountNotification(GCheckout.AutoGen.AuthorizationAmountNotification inputAuthorizationAmountNotification) {
    // TODO: Add custom processing for this notification type
  }

  private static void HandleChargeAmountNotification(GCheckout.AutoGen.ChargeAmountNotification inputChargeAmountNotification) {
    // TODO: Add custom processing for this notification type
  }

  private static void HandleNewOrderNotification(GCheckout.AutoGen.NewOrderNotification inputNewOrderNotification) {
    // Retrieve data from MerchantPrivateData
    GCheckout.AutoGen.anyMultiple oneAnyMultiple = inputNewOrderNotification.shoppingcart.merchantprivatedata;
    System.Xml.XmlNode[] oneXmlNodeArray = oneAnyMultiple.Any;
    string hiddenMerchantPrivateData = oneXmlNodeArray[0].InnerText;
    // TODO: Process the MerchantPrivateData if provided

    foreach (GCheckout.AutoGen.Item oneItem in inputNewOrderNotification.shoppingcart.items) {
      // TODO: Get MerchantItemId from shopping cart item (oneItem.merchantitemid) and process it
    }

    // TODO: Add custom processing for this notification type
  }

  private static void HandleOrderStateChangeNotification(GCheckout.AutoGen.OrderStateChangeNotification notification) {
    // Charge Order If Chargeable
    if ((notification.previousfinancialorderstate == GCheckout.AutoGen.FinancialOrderState.REVIEWING) 
      && (notification.newfinancialorderstate == GCheckout.AutoGen.FinancialOrderState.CHARGEABLE)) {
      
      ChargeOrderRequest oneChargeOrderRequest = new ChargeOrderRequest(notification.googleordernumber);
      GCheckout.Util.GCheckoutResponse oneGCheckoutResponse = oneChargeOrderRequest.Send();
    }

    // Update License If Charged
    if ((notification.previousfinancialorderstate == GCheckout.AutoGen.FinancialOrderState.CHARGING) 
      && (notification.newfinancialorderstate == GCheckout.AutoGen.FinancialOrderState.CHARGED)) {
      // TODO: For each shopping cart item received in the NewOrderNotification, authorize the license
    }

    // TODO: Add custom processing for this notification type
  }

  private static void HandleRiskInformationNotification(GCheckout.AutoGen.RiskInformationNotification notification) {
    // TODO: Add custom processing for this notification type
  }

  public static void ProcessNotification(string serialNumber) {

    //The next two statements set up a request and call google checkout for the details based on that serial number.

    NotificationHistoryRequest oneNotificationHistoryRequest
      = new NotificationHistoryRequest(new NotificationHistorySerialNumber(serialNumber));

    NotificationHistoryResponse oneNotificationHistoryResponse
      = (NotificationHistoryResponse)oneNotificationHistoryRequest.Send();

    // oneNotificationHistoryResponse.ResponseXml contains the complete response

    //what you need to do now is process the data that was returned.

    // Iterate through the notification history for this order looking for the notification that exactly matches the given serial number
    foreach (object oneNotification in oneNotificationHistoryResponse.NotificationResponses) {
      if (oneNotification.GetType().Equals(typeof(GCheckout.AutoGen.NewOrderNotification))) {
        GCheckout.AutoGen.NewOrderNotification oneNewOrderNotification = (GCheckout.AutoGen.NewOrderNotification)oneNotification;
        if (oneNewOrderNotification.serialnumber.Equals(serialNumber)) {
          HandleNewOrderNotification(oneNewOrderNotification);
        }
      }
      else if (oneNotification.GetType().Equals(typeof(GCheckout.AutoGen.OrderStateChangeNotification))) {
        GCheckout.AutoGen.OrderStateChangeNotification oneOrderStateChangeNotification = (GCheckout.AutoGen.OrderStateChangeNotification)oneNotification;
        if (oneOrderStateChangeNotification.serialnumber.Equals(serialNumber)) {
          HandleOrderStateChangeNotification(oneOrderStateChangeNotification);
        }
      }
      else if (oneNotification.GetType().Equals(typeof(GCheckout.AutoGen.RiskInformationNotification))) {
        GCheckout.AutoGen.RiskInformationNotification oneRiskInformationNotification = (GCheckout.AutoGen.RiskInformationNotification)oneNotification;
        if (oneRiskInformationNotification.serialnumber.Equals(serialNumber)) {
          HandleRiskInformationNotification(oneRiskInformationNotification);
        }
      }
      else if (oneNotification.GetType().Equals(typeof(GCheckout.AutoGen.AuthorizationAmountNotification))) {
        GCheckout.AutoGen.AuthorizationAmountNotification oneAuthorizationAmountNotification = (GCheckout.AutoGen.AuthorizationAmountNotification)oneNotification;
        if (oneAuthorizationAmountNotification.serialnumber.Equals(serialNumber)) {
          HandleAuthorizationAmountNotification(oneAuthorizationAmountNotification);
        }
      }
      else if (oneNotification.GetType().Equals(typeof(GCheckout.AutoGen.ChargeAmountNotification))) {
        GCheckout.AutoGen.ChargeAmountNotification oneChargeAmountNotification = (GCheckout.AutoGen.ChargeAmountNotification)oneNotification;
        if (oneChargeAmountNotification.serialnumber.Equals(serialNumber)) {
          HandleChargeAmountNotification(oneChargeAmountNotification);
        }
      }
      else {
        throw new ArgumentOutOfRangeException("Unhandled Type [" + oneNotification.GetType().ToString() + "]!; serialNumber=[" + serialNumber + "];");
      }
    }
  }
}
